### ISSUE TITLE
Wrong result when computing gradient in batches 

### ISSUE THREAD

**Issue by PhilipVinc** [FluxML/Zygote.jl#143]
I'm not sure Zygote is supposed to work in this case, but...

I have a function `f(x): R^N->R` and if I would like to evaluate the function for a batch of different inputs. This works fine where evaluating it such as `f(hcat(x1, x2))` but Zygote does not correctly batch the gradients.

```
using Zygote
f1(θ, x) = θ[:a]' * x
θ=Dict(:a=>rand(4), :w=>rand(4,4))
x=rand(4)
xb = hcat(x,x)

f1(θ, x)
> 0.41824058325741276
f1(θ, xb)
> 1×2 LinearAlgebra.Adjoint{Float64,Array{Float64,1}}:
>  0.418241  0.418241

val, back = Zygote.forward(θ->f1(θ, x), θ)
res = back([1 1])[1][:a]
> 4×1 LinearAlgebra.Adjoint{Float64,Base.ReshapedArray{Float64,2,LinearAlgebra.Transpose{Float> 64,Array{Float64,1}},Tuple{}}}:
>  0.00808089737288431
>  0.9809815233720736 
>  0.17521982530347224
>. 0.5098218410694737 

val, back = Zygote.forward(θ->f1(θ, xb), θ)
res = back([1 1])[1][:a]
> 4×1 LinearAlgebra.Adjoint{Float64,Array{Float64,2}}:
>  0.01616179474576862
>  1.9619630467441471 
>  0.3504396506069445 
>  1.0196436821389474 
```
Where I would expect him to return 
```
> 4×2 LinearAlgebra.Adjoint{Float64,Base.ReshapedArray{Float64,2,LinearAlgebra.Transpose{Float> 64,Array{Float64,1}},Tuple{}}}:
>  0.00808089737288431  0.00808089737288431
>  0.9809815233720736     0.9809815233720736 
>  0.17521982530347224    0.17521982530347224
>  0.5098218410694737      0.5098218410694737
```

**Comment by MikeInnes** [comment]
This is the intended behaviour, and any other AD (including Tracker) will behave the same way. If you sum your loss across a batch then the gradients will be the sum of gradients across the batch.

If you want something different you'd have to call `gradient` multiple times.

**Comment by PhilipVinc** [comment]
But I'm not summing my loss across a batch. When I call the loss function with a (batched) vector I get a vector with two outputs.

```
f1(θ, xb)
> 1×2 LinearAlgebra.Adjoint{Float64,Array{Float64,1}}:
>  0.418241  0.418241
```

**Comment by MikeInnes** [comment]
Yes, but you backpropagate the sensitivity `[1, 1]`, which equivalent to writing `sum(f1(..))`.

**Comment by PhilipVinc** [comment]
I see. So I should be simply back propagating `back(1)`?

**Comment by MikeInnes** [comment]
That won't do what you're looking for here. There's no way to get that behaviour automatically without something like Hydra.jl; you just have to write the loop.

**Comment by PhilipVinc** [comment]
Ok, so what I am trying to do is not supported at all.
Sorry for bothering you.

**Comment by MikeInnes** [comment]
No problem. Always worth asking for these kinds of things just in case :)
